File: Android/TRTCScenesDemo/trtcmeetingdemo/src/main/java/com/tencent/liteav/meeting/model/impl/room/impl/IMProtocol.java
package com.tencent.liteav.meeting.model.impl.room.impl;

import android.util.Pair;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.List;

import static com.tencent.liteav.meeting.model.impl.room.impl.IMProtocol.Define.*;

public class IMProtocol {

    public static class Define {
        public static final String KEY_VERSION = "version";
        public static final String KEY_ACTION = "action";
        public static final String VALUE_PROTOCOL_VERSION = "1.0.0";

        public static final int CODE_UNKNOWN                   = 0;

        public static final int CODE_ROOM_TEXT_MSG             = 300;
        public static final int CODE_ROOM_CUSTOM_MSG           = 301;
    }

    public static String getRoomTextMsgHeadJsonStr() {
        JSONObject jsonObject = new JSONObject();
        try {
            jsonObject.put(Define.KEY_VERSION, VALUE_PROTOCOL_VERSION);
            jsonObject.put(Define.KEY_ACTION, Define.CODE_ROOM_TEXT_MSG);
        } catch (JSONException e) {
            e.printStackTrace();
        }
        return jsonObject.toString();
    }

    public static String getCusMsgJsonStr(String cmd, String msg) {
        JSONObject jsonObject = new JSONObject();
        try {
            jsonObject.put(Define.KEY_VERSION, VALUE_PROTOCOL_VERSION);
            jsonObject.put(Define.KEY_ACTION, Define.CODE_ROOM_CUSTOM_MSG);
            jsonObject.put("command", cmd);
            jsonObject.put("message", msg);
        } catch (JSONException e) {
            e.printStackTrace();
        }
        return jsonObject.toString();
    }

    public static Pair<String, String> parseCusMsg(JSONObject jsonObject) {
        String cmd = jsonObject.optString("command");
        String message = jsonObject.optString("message");
        return new Pair<>(cmd, message);
    }
}

File: iOS/TRTCScenesDemo/TXLiteAVDemo/TRTCMeetingDemo/model/impl/IMProtocol.m
//
//  IMProtocol.m
//  TRTCScenesDemo
//
//  Created by J J on 2020/5/30.
//  Copyright © 2020 xcoderliu. All rights reserved.
//

#import "IMProtocol.h"

@implementation IMProtocol

+ (NSString *)getRoomTextMsgHeadJsonStr {
    NSMutableDictionary *dicJson = [[NSMutableDictionary alloc] init];
    [dicJson setValue:VALUE_PROTOCOL_VERSION forKey:KEY_VERSION];
    NSNumber *codeRoomTextMsg = [NSNumber numberWithInt:CODE_ROOM_TEXT_MESSAGE];
    [dicJson setValue:codeRoomTextMsg forKey:KEY_ACTION];
    
    return [IMProtocol convertToJsonData:dicJson];
}

+ (NSString *)getCusMsgJsonStr:(NSString *)cmd msg:(NSString *)msg {
    NSMutableDictionary *dicJson = [[NSMutableDictionary alloc] init];
    [dicJson setValue:VALUE_PROTOCOL_VERSION forKey:KEY_VERSION];
    [dicJson setValue:[NSNumber numberWithInt:CODE_ROOM_CUSTOM_MESSAGE] forKey:KEY_ACTION];
    [dicJson setValue:cmd forKey:@"command"];
    [dicJson setValue:msg forKey:@"message"];
    
    return [IMProtocol convertToJsonData:dicJson];
}

+ (TXMeetingPair *)parseCusMsg:(NSDictionary *)jsonObject {
    NSString *cmd = [jsonObject valueForKey:@"command"];
    NSString *message = [jsonObject valueForKey:@"message"];
    TXMeetingPair *pair = [[TXMeetingPair alloc] init];
    pair.first = cmd;
    pair.second = message;
    return pair;
}

+ (NSString *)convertToJsonData:(NSDictionary *)dict {
    NSError *error = NULL;
    NSData *jsonData = [NSJSONSerialization dataWithJSONObject:dict options:NSJSONWritingPrettyPrinted error:&error];
    if (error) {
        NSLog(@"%@", error);
        return @"";
    }
    return [[NSString alloc] initWithData:jsonData encoding:NSUTF8StringEncoding];
}

@end
